src/gz/write.rs

```diff
@@ -128,10 +128,8 @@ impl<W: Write> GzEncoder<W> {
     }
 
     fn write_header(&mut self) -> io::Result<()> {
-        while !self.header.is_empty() {
-            let n = self.inner.get_mut().write(&self.header)?;
-            self.header.drain(..n);
-        }
```

Comment:
This change we should probably avoid for safety - in practice I'd expect the entire header to written anyway, but as a corner case, this fine-grained write handling is certainly preferable.

Reply:
`write_all` makes sense here. It is essentially the same as the loop. However the drain or reassignment of `self.header` is not necessary. We can just leave that for the final deallocation of the encoder.

Reply:
Thanks for sharing!
I think the original concern comes from what happens if the write fails. Semantically, the current version would be able to continue exactly where it left off, when seen just by itself. I don't know if that is anything to be concerned about though, or if it's anything that would happen in practice.

Since you suggested to do the header draining/clearing on drop essentially, you don't seem to share the concern so maybe it's save to just go with the most obvious implementation?

But if so, why would someone write the current version in the first place, if not with the intend that it is granular.

Reply:
So it appears that the code is written this way to allow a `write` to fail with an error, but the caller can call `write` again and, if the error does not re-occur, then the file gets written correctly.

That might happen for non-blocking writes where EAGAIN gets returned. `write_all` isn't so useful in this case since it loses how many bytes were written before the EAGAIN (or an EINTR).

I agree then that it is safer to keep this as it is.